site/source/docs/porting/asyncify.rst

```diff
@@ -4,123 +4,220 @@
 
+.. note:: It's very important to optimize (``-O3`` here) when using Asyncify, as
+          unoptimized builds are very large.
 
-A hand written counterpart in JavaScript would be
+And you can run it with
 
 ::
 
-    function main() {
-      var i = 100;
-      console.log('Hello');
-      setTimeout(function() {
-        console.log('World ' + i + '!');
-        async_return_value = 0;
-      }, 1000);
-    }
+    nodejs a.out.js
 
-Specifically, a number of aspects should be taken into consideration:
- - Split the function when an async function is called, and the second function should be registered as the callback for the async function
- - Any function that calls an async function also becomes an async function.
- - Keep all local variables available to the callback
- - Closure cannot be used in order to make asm.js validated code.
- - Take care of loops and branches
- - Make the return value available to the callee
- - Some functions could be both sync or async, depending on the input.
+You should then see something like this:
 
-And the ``ASYNCIFY`` option does all above automatically, through a number of transformations on LLVM IR.
+::
 
+    sleeping...
+    sleeping...
+    sleeping...
+    sleeping...
+    sleeping...
+    timer happened!
 
-Usage
-=====
+The code is written with a straightforward loop, which does not exit while
+it is running, which normally would not allow async events to be handled by the
+browser. With Asyncify, those sleeps actually yield to the browser's main event
+loop, and the timer can happen!
 
-Call ``emscripten_sleep()`` whenever you need to pause the program, and add ``-s ASYNCIFY=1`` to emscripten.
+Making async Web APIs behave as if they were synchronous
+########################################################
 
-Sometimes it's a good replacement of ``emscripten_set_main_loop``, you may replace all ``sleep``-alike functions with ``emscripten_sleep``, instead of refactoring the whole main loop.
+Aside from ``emscripten_sleep`` and the other standard sync APIs Asyncify
+supports, you can also add your own functions. To do so, you must create a JS
+function that is called from wasm (since Emscripten controls pausing and
+resuming the wasm from the JS runtime). One way to do that is with a JS library
+function; another is to use ``EM_JS``, which we'll use in this next example:
 
-Also ``emscripten_sleep(1)`` can be used to 'interrupt' your code, such that the JavaScript engine can do the rendering and process events.
+::
 
-Extensions
-==========
+    // example.c
+    #include <emscripten.h>
+    #include <stdio.h>
 
-It is possible to implement more new async functions that appears to be sync in C.
+    EM_JS(void, do_fetch, (), {
+      Asyncify.handleSleep(function(wakeUp) {
+        out("waiting for a fetch");
+        fetch('a.html').then(response => {
+          out("got the fetch response");
+          // (normally you would do something with the fetch here)
```

Comment:
I guess if you wanted to get fancy, you could maybe demonstrate writing the data into memory or something but maybe that's beyond the scope of this example.